File: packages/go/messages.go
package rdk

// Type URLs for the qorechain.rdk.v1 transaction messages.
const rdkTypeURLPrefix = "/qorechain.rdk.v1."

const (
	TypeURLMsgCreateRollup      = rdkTypeURLPrefix + "MsgCreateRollup"
	TypeURLMsgSubmitBatch       = rdkTypeURLPrefix + "MsgSubmitBatch"
	TypeURLMsgChallengeBatch    = rdkTypeURLPrefix + "MsgChallengeBatch"
	TypeURLMsgResolveChallenge  = rdkTypeURLPrefix + "MsgResolveChallenge"
	TypeURLMsgPauseRollup       = rdkTypeURLPrefix + "MsgPauseRollup"
	TypeURLMsgResumeRollup      = rdkTypeURLPrefix + "MsgResumeRollup"
	TypeURLMsgStopRollup        = rdkTypeURLPrefix + "MsgStopRollup"
	TypeURLMsgExecuteWithdrawal = rdkTypeURLPrefix + "MsgExecuteWithdrawal"
)

// Msg is implemented by every rdk transaction message. TypeURL returns the wire
// type URL and Marshal returns the protobuf-encoded message bytes.
type Msg interface {
	TypeURL() string
	Marshal() []byte
}

// MsgCreateRollup registers a new rollup.
type MsgCreateRollup struct {
	Creator     string
	RollupID    string
	Profile     string
	VmType      string
	StakeAmount int64
}

func (MsgCreateRollup) TypeURL() string { return TypeURLMsgCreateRollup }

func (m MsgCreateRollup) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Creator)
	w.writeString(2, m.RollupID)
	w.writeString(3, m.Profile)
	w.writeString(4, m.VmType)
	w.writeInt64(5, m.StakeAmount)
	return w.bytes()
}

// MsgSubmitBatch submits a settlement batch.
type MsgSubmitBatch struct {
	Sequencer       string
	RollupID        string
	BatchIndex      uint64
	StateRoot       []byte
	PrevStateRoot   []byte
	TxCount         uint64
	DataHash        []byte
	Proof           []byte
	WithdrawalsRoot []byte
}

func (MsgSubmitBatch) TypeURL() string { return TypeURLMsgSubmitBatch }

func (m MsgSubmitBatch) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Sequencer)
	w.writeString(2, m.RollupID)
	w.writeUint64(3, m.BatchIndex)
	w.writeBytes(4, m.StateRoot)
	w.writeBytes(5, m.PrevStateRoot)
	w.writeUint64(6, m.TxCount)
	w.writeBytes(7, m.DataHash)
	w.writeBytes(8, m.Proof)
	w.writeBytes(9, m.WithdrawalsRoot)
	return w.bytes()
}

// MsgChallengeBatch challenges an optimistic batch with a fraud proof.
type MsgChallengeBatch struct {
	Challenger string
	RollupID   string
	BatchIndex uint64
	Proof      []byte
}

func (MsgChallengeBatch) TypeURL() string { return TypeURLMsgChallengeBatch }

func (m MsgChallengeBatch) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Challenger)
	w.writeString(2, m.RollupID)
	w.writeUint64(3, m.BatchIndex)
	w.writeBytes(4, m.Proof)
	return w.bytes()
}

// MsgResolveChallenge resolves an open challenge (upheld or dismissed).
type MsgResolveChallenge struct {
	Resolver    string
	RollupID    string
	BatchIndex  uint64
	FraudUpheld bool
}

func (MsgResolveChallenge) TypeURL() string { return TypeURLMsgResolveChallenge }

func (m MsgResolveChallenge) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Resolver)
	w.writeString(2, m.RollupID)
	w.writeUint64(3, m.BatchIndex)
	w.writeBool(4, m.FraudUpheld)
	return w.bytes()
}

// MsgPauseRollup pauses an active rollup.
type MsgPauseRollup struct {
	Creator  string
	RollupID string
	Reason   string
}

func (MsgPauseRollup) TypeURL() string { return TypeURLMsgPauseRollup }

func (m MsgPauseRollup) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Creator)
	w.writeString(2, m.RollupID)
	w.writeString(3, m.Reason)
	return w.bytes()
}

// MsgResumeRollup resumes a paused rollup.
type MsgResumeRollup struct {
	Creator  string
	RollupID string
}

func (MsgResumeRollup) TypeURL() string { return TypeURLMsgResumeRollup }

func (m MsgResumeRollup) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Creator)
	w.writeString(2, m.RollupID)
	return w.bytes()
}

// MsgStopRollup stops a rollup permanently.
type MsgStopRollup struct {
	Creator  string
	RollupID string
}

func (MsgStopRollup) TypeURL() string { return TypeURLMsgStopRollup }

func (m MsgStopRollup) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Creator)
	w.writeString(2, m.RollupID)
	return w.bytes()
}

// MsgExecuteWithdrawal finalizes an L2->L1 cross-layer message (withdrawal).
type MsgExecuteWithdrawal struct {
	Submitter       string
	RollupID        string
	BatchIndex      uint64
	WithdrawalIndex uint64
	Recipient       string
	Denom           string
	Amount          int64
	Proof           [][]byte
}

func (MsgExecuteWithdrawal) TypeURL() string { return TypeURLMsgExecuteWithdrawal }

func (m MsgExecuteWithdrawal) Marshal() []byte {
	w := &protoWriter{}
	w.writeString(1, m.Submitter)
	w.writeString(2, m.RollupID)
	w.writeUint64(3, m.BatchIndex)
	w.writeUint64(4, m.WithdrawalIndex)
	w.writeString(5, m.Recipient)
	w.writeString(6, m.Denom)
	w.writeInt64(7, m.Amount)
	w.writeRepeatedBytes(8, m.Proof)
	return w.bytes()
}

// --- Friendly message-builder inputs (mirror the TypeScript messages.ts) ---

// CreateRollupInput builds a MsgCreateRollup.
type CreateRollupInput struct {
	Creator     string
	RollupID    string
	Profile     string
	VmType      string
	StakeAmount int64
}

// CreateRollupMsg builds a MsgCreateRollup from input.
func CreateRollupMsg(in CreateRollupInput) MsgCreateRollup {
	return MsgCreateRollup{
		Creator:     in.Creator,
		RollupID:    in.RollupID,
		Profile:     in.Profile,
		VmType:      in.VmType,
		StakeAmount: in.StakeAmount,
	}
}

// SubmitBatchInput builds a MsgSubmitBatch.
type SubmitBatchInput struct {
	Sequencer       string
	RollupID        string
	BatchIndex      uint64
	StateRoot       []byte
	PrevStateRoot   []byte
	TxCount         uint64
	DataHash        []byte
	Proof           []byte
	WithdrawalsRoot []byte
}

// SubmitBatchMsg builds a MsgSubmitBatch from input.
func SubmitBatchMsg(in SubmitBatchInput) MsgSubmitBatch {
	return MsgSubmitBatch{
		Sequencer:       in.Sequencer,
		RollupID:        in.RollupID,
		BatchIndex:      in.BatchIndex,
		StateRoot:       in.StateRoot,
		PrevStateRoot:   in.PrevStateRoot,
		TxCount:         in.TxCount,
		DataHash:        in.DataHash,
		Proof:           in.Proof,
		WithdrawalsRoot: in.WithdrawalsRoot,
	}
}

// ChallengeBatchInput builds a MsgChallengeBatch.
type ChallengeBatchInput struct {
	Challenger string
	RollupID   string
	BatchIndex uint64
	Proof      []byte
}

// ChallengeBatchMsg builds a MsgChallengeBatch from input.
func ChallengeBatchMsg(in ChallengeBatchInput) MsgChallengeBatch {
	return MsgChallengeBatch(in)
}

// ResolveChallengeInput builds a MsgResolveChallenge.
type ResolveChallengeInput struct {
	Resolver    string
	RollupID    string
	BatchIndex  uint64
	FraudUpheld bool
}

// ResolveChallengeMsg builds a MsgResolveChallenge from input.
func ResolveChallengeMsg(in ResolveChallengeInput) MsgResolveChallenge {
	return MsgResolveChallenge(in)
}

// PauseRollupInput builds a MsgPauseRollup.
type PauseRollupInput struct {
	Creator  string
	RollupID string
	Reason   string
}

// PauseRollupMsg builds a MsgPauseRollup from input.
func PauseRollupMsg(in PauseRollupInput) MsgPauseRollup {
	return MsgPauseRollup(in)
}

// RollupRefInput identifies a rollup for resume/stop.
type RollupRefInput struct {
	Creator  string
	RollupID string
}

// ResumeRollupMsg builds a MsgResumeRollup from input.
func ResumeRollupMsg(in RollupRefInput) MsgResumeRollup {
	return MsgResumeRollup(in)
}

// StopRollupMsg builds a MsgStopRollup from input.
func StopRollupMsg(in RollupRefInput) MsgStopRollup {
	return MsgStopRollup(in)
}

// ExecuteWithdrawalInput builds a MsgExecuteWithdrawal.
type ExecuteWithdrawalInput struct {
	Submitter       string
	RollupID        string
	BatchIndex      uint64
	WithdrawalIndex uint64
	Recipient       string
	Denom           string
	Amount          int64
	Proof           [][]byte
}

// ExecuteWithdrawalMsg builds a MsgExecuteWithdrawal from input.
func ExecuteWithdrawalMsg(in ExecuteWithdrawalInput) MsgExecuteWithdrawal {
	return MsgExecuteWithdrawal(in)
}

File: packages/jvm/src/main/java/io/github/qorechain/rdk/tx/Messages.java
package io.github.qorechain.rdk.tx;

import io.github.qorechain.rdk.util.Bytes;
import java.util.ArrayList;
import java.util.List;

/**
 * The eight {@code qorechain.rdk.v1} transaction messages, hand-encoded to protobuf, plus friendly
 * input builders that mirror the reference clients. Field numbers and wire layout match the on-chain
 * {@code tx.proto} exactly.
 */
public final class Messages {
    private Messages() {}

    private static final String TYPE_URL_PREFIX = "/qorechain.rdk.v1.";

    public static final String TYPE_URL_MSG_CREATE_ROLLUP = TYPE_URL_PREFIX + "MsgCreateRollup";
    public static final String TYPE_URL_MSG_SUBMIT_BATCH = TYPE_URL_PREFIX + "MsgSubmitBatch";
    public static final String TYPE_URL_MSG_CHALLENGE_BATCH = TYPE_URL_PREFIX + "MsgChallengeBatch";
    public static final String TYPE_URL_MSG_RESOLVE_CHALLENGE = TYPE_URL_PREFIX + "MsgResolveChallenge";
    public static final String TYPE_URL_MSG_PAUSE_ROLLUP = TYPE_URL_PREFIX + "MsgPauseRollup";
    public static final String TYPE_URL_MSG_RESUME_ROLLUP = TYPE_URL_PREFIX + "MsgResumeRollup";
    public static final String TYPE_URL_MSG_STOP_ROLLUP = TYPE_URL_PREFIX + "MsgStopRollup";
    public static final String TYPE_URL_MSG_EXECUTE_WITHDRAWAL = TYPE_URL_PREFIX + "MsgExecuteWithdrawal";

    private static byte[] empty() {
        return new byte[0];
    }

    private static byte[] orEmpty(byte[] b) {
        return b == null ? empty() : b;
    }

    // --- Messages ---

    /** Registers a new rollup. */
    public static final class MsgCreateRollup implements Msg {
        public String creator = "";
        public String rollupId = "";
        public String profile = "";
        public String vmType = "";
        public long stakeAmount = 0;

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_CREATE_ROLLUP;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, creator);
            w.writeString(2, rollupId);
            w.writeString(3, profile);
            w.writeString(4, vmType);
            w.writeInt64(5, stakeAmount);
            return w.toByteArray();
        }
    }

    /** Submits a settlement batch. */
    public static final class MsgSubmitBatch implements Msg {
        public String sequencer = "";
        public String rollupId = "";
        public long batchIndex = 0;
        public byte[] stateRoot = empty();
        public byte[] prevStateRoot = empty();
        public long txCount = 0;
        public byte[] dataHash = empty();
        public byte[] proof = empty();
        public byte[] withdrawalsRoot = empty();

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_SUBMIT_BATCH;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, sequencer);
            w.writeString(2, rollupId);
            w.writeUint64(3, batchIndex);
            w.writeBytes(4, stateRoot);
            w.writeBytes(5, prevStateRoot);
            w.writeUint64(6, txCount);
            w.writeBytes(7, dataHash);
            w.writeBytes(8, proof);
            w.writeBytes(9, withdrawalsRoot);
            return w.toByteArray();
        }
    }

    /** Challenges an optimistic batch with a fraud proof. */
    public static final class MsgChallengeBatch implements Msg {
        public String challenger = "";
        public String rollupId = "";
        public long batchIndex = 0;
        public byte[] proof = empty();

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_CHALLENGE_BATCH;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, challenger);
            w.writeString(2, rollupId);
            w.writeUint64(3, batchIndex);
            w.writeBytes(4, proof);
            return w.toByteArray();
        }
    }

    /** Resolves an open challenge (upheld or dismissed). */
    public static final class MsgResolveChallenge implements Msg {
        public String resolver = "";
        public String rollupId = "";
        public long batchIndex = 0;
        public boolean fraudUpheld = false;

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_RESOLVE_CHALLENGE;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, resolver);
            w.writeString(2, rollupId);
            w.writeUint64(3, batchIndex);
            w.writeBool(4, fraudUpheld);
            return w.toByteArray();
        }
    }

    /** Pauses an active rollup. */
    public static final class MsgPauseRollup implements Msg {
        public String creator = "";
        public String rollupId = "";
        public String reason = "";

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_PAUSE_ROLLUP;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, creator);
            w.writeString(2, rollupId);
            w.writeString(3, reason);
            return w.toByteArray();
        }
    }

    /** Resumes a paused rollup. */
    public static final class MsgResumeRollup implements Msg {
        public String creator = "";
        public String rollupId = "";

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_RESUME_ROLLUP;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, creator);
            w.writeString(2, rollupId);
            return w.toByteArray();
        }
    }

    /** Stops a rollup permanently. */
    public static final class MsgStopRollup implements Msg {
        public String creator = "";
        public String rollupId = "";

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_STOP_ROLLUP;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, creator);
            w.writeString(2, rollupId);
            return w.toByteArray();
        }
    }

    /** Finalizes an L2→L1 cross-layer message (withdrawal). */
    public static final class MsgExecuteWithdrawal implements Msg {
        public String submitter = "";
        public String rollupId = "";
        public long batchIndex = 0;
        public long withdrawalIndex = 0;
        public String recipient = "";
        public String denom = "";
        public long amount = 0;
        public List<byte[]> proof = new ArrayList<>();

        @Override
        public String typeUrl() {
            return TYPE_URL_MSG_EXECUTE_WITHDRAWAL;
        }

        @Override
        public byte[] marshal() {
            ProtoWriter w = new ProtoWriter();
            w.writeString(1, submitter);
            w.writeString(2, rollupId);
            w.writeUint64(3, batchIndex);
            w.writeUint64(4, withdrawalIndex);
            w.writeString(5, recipient);
            w.writeString(6, denom);
            w.writeInt64(7, amount);
            w.writeRepeatedBytes(8, proof);
            return w.toByteArray();
        }
    }

    // --- Friendly input builders ---

    public static final class CreateRollupInput {
        public String creator;
        public String rollupId;
        public String profile;
        public String vmType;
        public long stakeAmount;
    }

    public static MsgCreateRollup createRollupMsg(CreateRollupInput in) {
        MsgCreateRollup m = new MsgCreateRollup();
        m.creator = in.creator;
        m.rollupId = in.rollupId;
        m.profile = in.profile;
        m.vmType = in.vmType;
        m.stakeAmount = in.stakeAmount;
        return m;
    }

    public static final class SubmitBatchInput {
        public String sequencer;
        public String rollupId;
        public long batchIndex;
        public byte[] stateRoot;
        public byte[] prevStateRoot;
        public long txCount;
        public byte[] dataHash;
        public byte[] proof;
        public byte[] withdrawalsRoot;
    }

    public static MsgSubmitBatch submitBatchMsg(SubmitBatchInput in) {
        MsgSubmitBatch m = new MsgSubmitBatch();
        m.sequencer = in.sequencer;
        m.rollupId = in.rollupId;
        m.batchIndex = in.batchIndex;
        m.stateRoot = orEmpty(in.stateRoot);
        m.prevStateRoot = orEmpty(in.prevStateRoot);
        m.txCount = in.txCount;
        m.dataHash = orEmpty(in.dataHash);
        m.proof = orEmpty(in.proof);
        m.withdrawalsRoot = orEmpty(in.withdrawalsRoot);
        return m;
    }

    public static final class ChallengeBatchInput {
        public String challenger;
        public String rollupId;
        public long batchIndex;
        public byte[] proof;
    }

    public static MsgChallengeBatch challengeBatchMsg(ChallengeBatchInput in) {
        MsgChallengeBatch m = new MsgChallengeBatch();
        m.challenger = in.challenger;
        m.rollupId = in.rollupId;
        m.batchIndex = in.batchIndex;
        m.proof = orEmpty(in.proof);
        return m;
    }

    public static final class ResolveChallengeInput {
        public String resolver;
        public String rollupId;
        public long batchIndex;
        public boolean fraudUpheld;
    }

    public static MsgResolveChallenge resolveChallengeMsg(ResolveChallengeInput in) {
        MsgResolveChallenge m = new MsgResolveChallenge();
        m.resolver = in.resolver;
        m.rollupId = in.rollupId;
        m.batchIndex = in.batchIndex;
        m.fraudUpheld = in.fraudUpheld;
        return m;
    }

    public static final class PauseRollupInput {
        public String creator;
        public String rollupId;
        public String reason;
    }

    public static MsgPauseRollup pauseRollupMsg(PauseRollupInput in) {
        MsgPauseRollup m = new MsgPauseRollup();
        m.creator = in.creator;
        m.rollupId = in.rollupId;
        m.reason = in.reason == null ? "" : in.reason;
        return m;
    }

    public static final class RollupRefInput {
        public String creator;
        public String rollupId;
    }

    public static MsgResumeRollup resumeRollupMsg(RollupRefInput in) {
        MsgResumeRollup m = new MsgResumeRollup();
        m.creator = in.creator;
        m.rollupId = in.rollupId;
        return m;
    }

    public static MsgStopRollup stopRollupMsg(RollupRefInput in) {
        MsgStopRollup m = new MsgStopRollup();
        m.creator = in.creator;
        m.rollupId = in.rollupId;
        return m;
    }

    public static final class ExecuteWithdrawalInput {
        public String submitter;
        public String rollupId;
        public long batchIndex;
        public long withdrawalIndex;
        public String recipient;
        public String denom;
        public long amount;
        public List<byte[]> proof;
    }

    public static MsgExecuteWithdrawal executeWithdrawalMsg(ExecuteWithdrawalInput in) {
        MsgExecuteWithdrawal m = new MsgExecuteWithdrawal();
        m.submitter = in.submitter;
        m.rollupId = in.rollupId;
        m.batchIndex = in.batchIndex;
        m.withdrawalIndex = in.withdrawalIndex;
        m.recipient = in.recipient;
        m.denom = in.denom;
        m.amount = in.amount;
        m.proof = in.proof == null ? new ArrayList<>() : in.proof;
        return m;
    }

    /** Coerce a hex string into bytes (convenience for callers building byte fields). */
    public static byte[] bytes(String hex) {
        return Bytes.toBytes(hex);
    }
}
